activemodel/lib/active_model/type/helpers/mutable.rb

```diff
@@ -4,6 +4,10 @@ module ActiveModel
   module Type
     module Helpers # :nodoc: all
       module Mutable
+        def immutable_value(value)
+          value.deep_dup.freeze
```

Comment:
@tenderlove : this is causing an issue with serialization (sorry, I'm not quite sure yet how to describe).
I am troubleshooting a RED CI that I started having after upgrading to Rails 7.0.5.1 (from 7.0.4.3) and I traced it down to this line.

Here is what is happening in my case:
I have a `p` var which is hash that is like:
```
[6] pry(#<AdminMailerPreview>)> p
=> {:business=>
  #<Business:0x000000010a3556c8
   id: 1,
   user_id: 34,
   contact_name: "Rufus Jenkins",
   company: "Business Company",
   created_at: Wed, 12 Jul 2023 08:22:23.573859000 UTC +00:00,
   updated_at: Wed, 12 Jul 2023 08:22:26.052893000 UTC +00:00,
   bio:
    "Nihil qui autem. Porro animi repellendus. Officia reprehenderit saepe. Placeat accusantium consequatur. Aut distinctio similique. Placeat et dolores. Ex quibusdam dicta. Quia eaque ut. Ut ut magnam. Eos explicabo alias.",
   developer_notifications: "daily",
   website: nil,
   contact_role: nil,
   invisible: false,
   survey_request_notifications: true>}
```

When I run this line inside my test, here is the result I get:
```
[11] pry(#<AdminMailerPreview>)> p.deep_dup.freeze
=> {:business=>
  #<Business:0x000000010b0b0b60
   id: nil,
   user_id: 34,
   contact_name: "Rufus Jenkins",
   company: "Business Company",
   created_at: nil,
   updated_at: nil,
   bio:
    "Nihil qui autem. Porro animi repellendus. Officia reprehenderit saepe. Placeat accusantium consequatur. Aut distinctio similique. Placeat et dolores. Ex quibusdam dicta. Quia eaque ut. Ut ut magnam. Eos explicabo alias.",
   developer_notifications: "daily",
   website: nil,
   contact_role: nil,
   invisible: false,
   survey_request_notifications: true>}
```

After this, my test ends up failing here: https://github.com/rails/rails/blob/v7.0.6/activejob/lib/active_job/arguments.rb#L95

Because my new `p` hash has `id: nil`, I get an exception raised: `I::GID::MissingModelIdError: Unable to create a Global ID for Business without a model id.`, here: https://github.com/rails/rails/blob/v7.0.6/activejob/lib/active_job/arguments.rb#L204-L209

If I comment out this entire method, my CI tests pass.
If I remove the `deep_dup` (leaving this method returning `value.freeze`, my CI tests pass.

You can see the CI failure in here: https://github.com/joemasilotti/railsdevs.com/pull/874
This issue first started when upgrading from Rails 7.0.4.3 to 7.0.5.1.

Is there a reason why you must call `deep_dup` in the value object here? Or just the freeze suffice?
Tomorrow I will try to dig more into the issue in this PR and see if I can figure out a different way to solve the issue, that doesn't affect my use case.

Let me know if I can provide more details.
I'd be happy to open a PR and work on a solution! :blush:

FYI: this is the test that is failing: https://github.com/joemasilotti/railsdevs.com/blob/main/test/mailers/previews/admin_mailer_preview.rb#L8-L12

Reply:
Thank you for all of the details, there is already an issue open for this bug: #48652 and Jean is working on a PR to fix it as well: #48705

In the future please don't ping individual maintainers. They get plenty of notifications as it is and your message is much more likely to be seen if you just open an issue with reproduction steps